src/pdb/string.rs

```diff
@@ -0,0 +1,271 @@
+        }
+    }
+    /// Create an empty DeviceSQLString
+    ///
+    /// Should be used to construct known empty strings.
+    #[must_use]
+    pub const fn empty() -> Self {
+        Self(DeviceSQLStringImpl::ShortASCII {
+            content: Vec::new(),
+        })
+    }
+}
+
+/// A String encapsulating how DeviceSQL VARCHAR's are structured
+///
+/// This implementation forces them to be immutable for now
+#[binrw]
+#[derive(Debug, PartialEq)]
+enum DeviceSQLStringImpl {
+    /// Short-String optimization case
+    ShortASCII {
+        #[br(temp, assert(header & 0b1 == 1))]
+        #[bw(calc = (((content.len() + 1) << 1) | 1) as u8)]
+        header: u8,
+
+        #[br(count = (header >> 1) - 1)]
+        content: Vec<u8>,
+    },
+    /// Regular long form strings, containing possibly different encodings
+    Long {
+        #[br(temp)]
+        #[br(assert(flags & 0b1 == 0))]
+        #[bw(calc = content.flags())]
+        flags: u8,
+
+        #[br(temp)]
+        #[bw(calc = content.byte_count().unwrap() + 4)]
+        length: u16,
+
+        #[brw(magic(0u8))] // padding
+        #[br(args(flags, length - 4))]
+        content: LongBody,
+    },
+}
+
+#[binrw]
+#[derive(Debug, PartialEq)]
+#[br(import(flags: u8, len: u16))]
+enum LongBody {
+    // Ordering is important otherwise, UCS2LE strings could be parsed
+    // instead of the stricter ISRC
+
+    // ISRC strings are a bug/flaw in pioneers implementation, this is technically
+    // a semi-dirty workaround
+    #[br(pre_assert(flags == 0x90))]
+    Isrc(#[brw(magic = 0x3u8)] binrw::NullString),
+    #[br(pre_assert(flags == 0x40))]
+    Ascii(#[br(count = len)] Vec<u8>),
+    #[br(pre_assert(flags == 0x90))]
+    #[br(pre_assert(len % 2 == 0))]
+    Ucs2le(#[br(count = len / 2)] Vec<u16>),
+}
+
+impl LongBody {
+    pub fn byte_count(&self) -> Result<u16, StringError> {
+        match self {
+            // ISRC offset is compensating for trailing nullbyte + 0x3 magic byte.
+            Self::Isrc(null_str) => null_str.len() + 2,
+            Self::Ascii(buf) => buf.len(),
+            Self::Ucs2le(buf) => (buf.len() * 2),
+        }
+        .try_into()
+        .map_err(|_| StringError::TooLong)
+    }
+    pub fn flags(&self) -> u8 {
+        match self {
+            Self::Ucs2le(_) | Self::Isrc(_) => 0x90,
+            Self::Ascii(_) => 0x40,
+        }
+    }
+}
+
+impl Default for DeviceSQLString {
+    fn default() -> Self {
+        Self::empty()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::util::testing::test_roundtrip;
+
+    #[test]
+    fn default_string() -> Result<(), StringError> {
+        test_roundtrip(&[0x3], DeviceSQLString::default());
+        Ok(())
+    }
+
+    #[test]
+    fn short_ascii_string() -> Result<(), StringError> {
+        test_roundtrip(
+            &[0x9, 0x66, 0x6F, 0x6F],
+            DeviceSQLString::new("foo".to_owned())?,
+        );
+        Ok(())
+    }
```

Comment:
I'd also be okay with the more concise version, but I don't really care:

```suggestion
    #[test]
    fn short_ascii_string() {
        test_roundtrip(
            &[0x9, 0x66, 0x6F, 0x6F],
            DeviceSQLString::new("foo".to_owned()).unwrap(),
        );
    }
```